File: Core/Workspace/Typescript/libs/promise/core/src/Context.ts
import { Database, Session, Method } from '@allors/domain/system';
import { Pull } from '@allors/data/system';
import { PullRequest, PullResponse, SyncResponse, PushRequest, PushResponse, SyncRequest, PushRequestObject, InvokeResponse } from '@allors/protocol/system';

import { Client } from './Client';

import { Invoked } from './responses/Invoked';
import { Loaded } from './responses/Loaded';
import { Saved } from './responses/Saved';

export class Context {
  public session: Session;

  constructor(public client: Client, public database: Database) {
    this.session = database.createSession();
  }

  public load(pull: Pull | PullRequest): Promise<Loaded>;
  public load(customService: string, customArgs?: any): Promise<Loaded>;
  public load(requestOrCustomService: PullRequest | Pull | string, customArgs?: any): Promise<Loaded> {
    return new Promise((resolve, reject) => {
      this.client
        .pull(requestOrCustomService, customArgs)
        .then((pullResponse: PullResponse) => {
          const syncRequest = this.database.diff(pullResponse);
          if (syncRequest.objects.length > 0) {
            return this.client
              .sync(syncRequest)
              .then((syncResponse: SyncResponse) => {
                const securityRequest = this.database.sync(syncResponse);
                if (securityRequest) {
                  this.client
                    .security(securityRequest)
                    .then((v) => {
                      const securityRequest2 = this.database.security(v);
                      if (securityRequest2) {
                        this.client
                          .security(securityRequest2)
                          .then((v) => {
                            this.database.security(v);
                            const loaded = new Loaded(this.session, pullResponse);
                            resolve(loaded);
                          })
                          .catch((e) => {
                            reject(e);
                          });
                      } else {
                        const loaded = new Loaded(this.session, pullResponse);
                        resolve(loaded);
                      }
                    })
                    .catch((e) => {
                      reject(e);
                    });
                } else {
                  const loaded = new Loaded(this.session, pullResponse);
                  resolve(loaded);
                }
              })
              .catch((e) => {
                reject(e);
              });
          } else {
            const loaded = new Loaded(this.session, pullResponse);
            resolve(loaded);
            return;
          }
        })
        .catch((e) => {
          reject(e);
        });
    });
  }

  public save(): Promise<Saved> {
    return new Promise((resolve, reject) => {
      const pushRequest: PushRequest = this.session.pushRequest();
      return this.client
        .push(pushRequest)
        .then((pushResponse: PushResponse) => {
          this.session.pushResponse(pushResponse);
          const syncRequest: SyncRequest = {
            objects: pushRequest.objects?.map((v: PushRequestObject) => v.i) ?? []
          };

          if (pushResponse.newObjects) {
            for (const newObject of pushResponse.newObjects) {
              syncRequest.objects.push(newObject.i);
            }
          }

          return this.client
            .sync(syncRequest)
            .then((syncResponse: SyncResponse) => {
              this.database.sync(syncResponse);
              const saved: Saved = new Saved(this.session, pushResponse);
              resolve(saved);
            })
            .catch((e) => {
              reject(e);
            });
        })
        .catch((e) => {
          reject(e);
        });
    });
  }

  public invoke(method: Method): Promise<Invoked>;
  public invoke(service: string, args?: any): Promise<Invoked>;
  public invoke(methodOrService: Method | string, args?: any): Promise<Invoked> {
    return this.client
      .invoke(methodOrService as any, args)
      .then((invokeResponse: InvokeResponse) => new Invoked(this.session, invokeResponse));
  }
}

File: Platform/Workspace/CSharp/Adapters/Allors.Workspace.Adapters.Remote/Core/Workspace/Context.cs
// <copyright file="Context.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Remote
{
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading.Tasks;
    using Protocol.Database.Invoke;
    using Protocol.Database.Pull;
    using Protocol.Database.Push;
    using Protocol.Database.Sync;
    using Allors.Workspace.Data;
    using Result = Allors.Workspace.Result;

    public class Context : IContext
    {
        private readonly ContextFactory contextFactory;
        private readonly Workspace workspace;
        private readonly ClientDatabase database;

        public Context(ContextFactory contextFactory)
        {
            this.contextFactory = contextFactory;
            this.workspace = this.contextFactory.Workspace;
            this.database = this.contextFactory.Database;

            // TODO: remove cast
            this.Session = new Session((Workspace)this.workspace, this.workspace.Lifecycle.CreateSessionScope());

            this.contextFactory.RegisterContext(this);
        }

        ~Context() => this.contextFactory.UnregisterContext(this);

        ISession IContext.Session => this.Session;

        public Session Session { get; }

        public Task<InvokeResponse> Invoke(Method method, InvokeOptions options = null) => this.Invoke(new[] { method }, options);

        public Task<InvokeResponse> Invoke(Method[] methods, InvokeOptions options = null)
        {
            var invokeRequest = new InvokeRequest
            {
                I = methods.Select(v => new Invocation
                {
                    I = v.Object.Id.ToString(),
                    V = v.Object.Strategy.Version.ToString(),
                    M = v.MethodType.IdAsString,
                }).ToArray(),
                O = options,
            };

            return this.database.Invoke(invokeRequest);
        }

        public Task<InvokeResponse> Invoke(string service, object args) => this.database.Invoke(service, args);

        public async Task<Result> Load(params Pull[] pulls)
        {
            var pullRequest = new PullRequest { P = pulls.Select(v => v.ToJson()).ToArray() };
            var pullResponse = await this.database.Pull(pullRequest);
            var syncRequest = this.workspace.Diff(pullResponse);
            if (syncRequest.Objects.Length > 0)
            {
                await this.Load(syncRequest);
            }

            var result = new Result(this.Session, pullResponse);
            return result;
        }

        public async Task<Result> Load(object args, string pullService = null)
        {
            if (args is Pull pull)
            {
                args = new PullRequest { P = new[] { pull.ToJson() } };
            }

            if (args is IEnumerable<Pull> pulls)
            {
                args = new PullRequest { P = pulls.Select(v => v.ToJson()).ToArray() };
            }

            var pullResponse = await this.database.Pull(pullService, args);
            var syncRequest = this.workspace.Diff(pullResponse);

            if (syncRequest.Objects.Length > 0)
            {
                await this.Load(syncRequest);
            }

            var result = new Result(this.Session, pullResponse);
            return result;
        }

        public async Task<PushResponse> Save()
        {
            var saveRequest = this.Session.PushRequest();
            var pushResponse = await this.database.Push(saveRequest);
            if (!pushResponse.HasErrors)
            {
                this.Session.PushResponse(pushResponse);

                var objects = saveRequest.Objects.Select(v => v.I).ToArray();
                if (pushResponse.NewObjects != null)
                {
                    objects = objects.Union(pushResponse.NewObjects.Select(v => v.I)).ToArray();
                }

                var syncRequests = new SyncRequest
                {
                    Objects = objects,
                };

                await this.Load(syncRequests);

                this.Session.Reset();
            }

            return pushResponse;
        }

        private async Task Load(SyncRequest syncRequest)
        {
            var syncResponse = await this.database.Sync(syncRequest);
            var securityRequest = this.workspace.Sync(syncResponse);

            if (securityRequest != null)
            {
                var securityResponse = await this.database.Security(securityRequest);
                securityRequest = this.workspace.Security(securityResponse);

                if (securityRequest != null)
                {
                    securityResponse = await this.database.Security(securityRequest);
                    this.workspace.Security(securityResponse);
                }
            }
        }
    }
}
